### DISCUSSION TITLE
How to get OpenOCD de-bugger up and running

### DISCUSSION THREAD

**Discussion by DS-567** [stnolting/neorv32#488] [General]
I am trying to get OpenOCD debugger working with GBD according to the example in the NEORV32 user guide. I am running the demo_blink_led with the on-chip debugger test setup modified to map the lower 8-bits of the GPIO_o to 8 LEDs on the FPGA (BASYS 3 dev board), which is working perfect. 

I am using the FT2232H-56Q JTAG adapter as an interface between the NEORV32 and laptop with the following connections as described in the user guide:

jtag_tck_i    => CN2 D0
jtag_tdi_i    => CN2 D1
jtag_tdo_o  => CN2 D2
jtag_tms_i   => CN2 D3
jtag_trst_i    => CN2 D4

I have git cloned OpenOCD onto my Ubuntu directory and modified my path variable to suit (I think), as well as installed the drivers for the FTDI JTAG device.

With the FTDI plugged into the USB port and no connections to the NEORV32 JTAG I/O yet, when I try to run OpenOCD from Ubuntu, I get the following error: "Error: invalid command name "fdti". I have very limited experience with Ubuntu, but does this mean that the configuration file (in the NEORV32 code folder saved in Windows directory) cannot access the OpenOCD files stored on Ubuntu? Or any advice on how to fix this?  

![Capture](https://user-images.githubusercontent.com/100368901/217651007-6a055bc1-5666-4bf2-89d6-5a4ce58a3fa4.PNG)

**Comment by stnolting** [comment]
I have never seen that specific error. Maybe there is a problem with the drivers... You could turn on the advanced debugging output to see what's going on: `openocd -d3 -f openocd_neorv32.cfg`

Did you compile openOCD by yourself or did you install a pre-build version?

You could try installing openOCD using your package manager: `sudo apt install openocd`. Alternatively, you could manually install a pre-compiled version. I can recommend xpack: https://github.com/xpack-dev-tools/openocd-xpack/releases

**Reply by DS-567** [reply]
I rebuilt the openOCD install and compilation and now getting this error message (different error):

![image](https://user-images.githubusercontent.com/100368901/218182029-6265f9eb-3bd7-462b-a1d5-07b623dbdf55.png)

> Did you compile openOCD by yourself or did you install a pre-build version?

I have git cloned openOCD from here (https://github.com/riscv/riscv-openocd) and followed the guide in this tutorial (https://robbert-groen.medium.com/getting-started-with-the-gd32vf103-risc-v-microcontroller-22cb34718b0d)  under "Installing openOCD" with steps 1 to 5. Seemed to work ok.

Appears that the board is not reachable. Does it matter that I am running the openOCD from Ubuntu while I installed the FTDI drivers on Windows?

**Reply by stnolting** [reply]
> Appears that the board is not reachable. Does it matter that I am running the openOCD from Ubuntu while I installed the FTDI drivers on Windows?

That might be the problem...

Are you using a virtual machine on Windows? Or do you use the Windows Subsystem for Linux (WSL)?
I run openOCD under Windows and GDB runs inside Ubuntu/WSL. Might that be a suitbale setup for you?

**Reply by DS-567** [reply]
> I run openOCD under Windows and GDB runs inside Ubuntu/WSL. Might that be a suitbale setup for you?

Do I need to install openOCD on Windows then first?

**Reply by stnolting** [reply]
Right. I use the Windows version of openOCD. Luckily, Windows already has all the required FTDI drivers on board 😉

**Reply by stnolting** [reply]
Btw, this step-by-step guide is based on my setup: https://stnolting.github.io/neorv32/ug/#_openocd

**Reply by DS-567** [reply]
>  Right. I use the Windows version of openOCD. Luckily, Windows already has all the required FTDI drivers on board 😉

Alright, thanks, I will try that. Ok to install from here? : https://github.com/riscv/riscv-openocd

> Btw, this step-by-step guide is based on my setup: https://stnolting.github.io/neorv32/ug/#_openocd

Yeah, I am following through that example, just having issues with accessing the board through Ubuntu I think. Going to try on Windows now.

**Reply by stnolting** [reply]
I can recommend the xpack bundles: https://github.com/xpack-dev-tools/openocd-xpack/releases

**Reply by stnolting** [reply]
This seems fine for Windows: https://github.com/xpack-dev-tools/openocd-xpack/releases/tag/v0.11.0-5

**Reply by DS-567** [reply]
Ok, another different error message: 

![image](https://user-images.githubusercontent.com/100368901/218274565-4b733242-0464-446c-ac76-2e51c33eef8b.png)

Not sure what is wrong with this at all 😅. This must be a library not being found?

**Reply by stnolting** [reply]
😅
Can you connect to the FTDI adapter using a simple terminal program (like TeraTerm)?
Does your adapter show up in the Windows device manger (and without any warnings)?
Also check if the adapter actually uses the lib-usb driver.

**Reply by DS-567** [reply]
> Can you connect to the FTDI adapter using a simple terminal program (like TeraTerm)?

Yes TeraTerm recognises the device connected to com port 3 and 4.

> Does your adapter show up in the Windows device manger (and without any warnings)?

Also it can be seen in the device manager no problem.

If it makes a difference I have the board plugged into the USB port with both jumpers connected and no other connections on the board. Does the reset pin matter at this stage?

**Reply by stnolting** [reply]
> Yes TeraTerm recognises the device connected to com port 3 and 4.

I think this is just a driver issue... The FTDI provides two channels (A and B). In your current configuration, both are configured as virtual COM ports using the **default FTDI driver** - that is why you get two COM ports in TeraTerm. You need to reconfigure the chip's JTAG port to use **libusb** instead.

This is what my driver configuration looks like:
![grafik](https://user-images.githubusercontent.com/22944758/218518208-f53fe3e2-dcaf-4b4d-92c5-9394df2b8ee6.png)

In TeraTerm, I see just one virtual COM port.

You can use the USB-Driver-Tool from https://visualgdb.com/UsbDriverTool/ to reconfigure each port with the required driver.

**Reply by DS-567** [reply]
Ok we are maybe getting somewhere now. Here are port driver configurations now:

![image](https://user-images.githubusercontent.com/100368901/218520270-e45de7c1-6776-4e7a-aad0-ff3af43ce9af.png)

And here is the cmd terminal output:

![image](https://user-images.githubusercontent.com/100368901/218520418-ddba8400-fb47-43b4-9803-8e2a03665fb6.png)

Is this more like what is expected without the NEORV32 connected to the FT2232H?

**Reply by stnolting** [reply]
> Ok we are maybe getting somewhere now. Here are port driver configurations now:

That looks good!

> And here is the cmd terminal output:

That also looks good - at least the driver is working now.

> Is this more like what is expected without the NEORV32 connected to the FT2232H?

Right. This is what you get if your FPGA is not configured yet (or if the FTDI-FPGA wiring is incorrect).

**Reply by DS-567** [reply]
Ok, so to install and run GDB side of the on-chip debugger, do you recommend doing this on Ubuntu over Windows

**Reply by stnolting** [reply]
I am using Ubuntu (WSL) for all the compiler-related things. But this is a matter of taste I think 😉

**Reply by DS-567** [reply]
Sorry for being a pain, my software skills are very limited 😅. Following the guide in the user guide, I am now getthing the following error message:

![image](https://user-images.githubusercontent.com/100368901/218554680-244bce5a-0704-4946-99a3-a2bbe8363ff8.png)

**Reply by stnolting** [reply]
No worries! I was in the same situation about two years ago when I had absolutely no clue about GDB and all that (I am no software guy at all) :wink:

First of all, can openOCD connect to the processor?

> Following the guide in the user guide, I am now getthing the following error message:

What kind of toolchain are you using? Can you run `make info` and post the result (or just the "Info: Compiler Configuration" section)?

I am not quite sure why a Python library is missing or even required here... I would try updating your system first (`sudo apt update` and `sudo apt upgrade`).

**Reply by DS-567** [reply]
> First of all, can openOCD connect to the processor?

Yes it can connect (see below image) although that libusb() library not found error message still shows. This probably won't affect debugging I guess?

![image](https://user-images.githubusercontent.com/100368901/218689757-6435bffe-e4c9-4a9a-8502-d50d53db47bc.png)

> What kind of toolchain are you using?

I am using the pre-built recommended (rv32i-4.0.0) in the user guide from here (https://github.com/stnolting/riscv-gcc-prebuilt). I assume this includes the installation for GDB?

It seems to be installed ok?

![image](https://user-images.githubusercontent.com/100368901/218728887-37f31b9e-f249-439d-81a0-0844972b79ac.png)


> I am not quite sure why a Python library is missing or even required here... I would try updating your system first (sudo apt update and sudo apt upgrade).

I did try this first but no joy

**Reply by stnolting** [reply]
> Yes it can connect (see below image) although that libusb() library not found error message still shows. This probably won't affect debugging I guess?

This is strange. But it seems to work - so let's ignore this for now 😅

> I am using the pre-built recommended (rv32i-4.0.0) in the user guide from here (https://github.com/stnolting/riscv-gcc-prebuilt). I assume this includes the installation for GDB?

Great. GDB is included in that toolchain.

> It seems to be installed ok?

I think you are calling the **native** GDB of your system here.

Let's first check the toolchain installation in terms of the NEORV32 project. Please naviagte with your console to one of the project's example program and run the toolchain check:

```bash
neorv32/sw/example/demo_blink_led$ make check
```

If "Toolchain check OK" appears at the end then the installation if fine.

In that case please start the GDB version **from the RISC-V toolchain**:

```bash
$ riscv32-unknown-elf-gdb
```

All these steps require that you have added the toolchain's `bin` folder to your `PATH` environment variable.

**Reply by DS-567** [reply]
> Let's first check the toolchain installation in terms of the NEORV32 project. Please naviagte with your console to one of the project's example program and run the toolchain check:

`neorv32/sw/example/demo_blink_led$ make check`

> If "Toolchain check OK" appears at the end then the installation if fine.

Seems to be installed correctly (see screenshots of the Ubuntu terminal). I have added the path: `$ export PATH=$PATH:/opt/riscv/bin` 

![image](https://user-images.githubusercontent.com/100368901/218988031-21ea24c1-4f4f-40d8-b82e-60df8e92d1e7.png)

![image](https://user-images.githubusercontent.com/100368901/218988198-52ca13ee-f6a4-4280-bcb7-f78a390a2eaa.png)

Same error when running this: `$ riscv32-unknown-elf-gdb`

![image](https://user-images.githubusercontent.com/100368901/218554680-244bce5a-0704-4946-99a3-a2bbe8363ff8.png)

Maybe I need Python installed on Ubuntu?

**Reply by stnolting** [reply]
> Maybe I need Python installed on Ubuntu?

Seems like the compiler requires Python to be installed. So maybe it is worth a try. 😉

**Reply by DS-567** [reply]
I actually already have Python 3.10 already installed on Ubuntu, so will try with Python 3.8.

**Reply by stnolting** [reply]
Sounds good!

**Reply by DS-567** [reply]
Finally I think it now seems to be working 😅😅😅 :

![image](https://user-images.githubusercontent.com/100368901/219096092-29916928-0b9e-4d51-8842-391876710991.png)

For anyone facing the same issue, I had to install Python 3.8 on Ubuntu then run the following command:

`sudo ln -s /usr/lib/x86_64-linux-gnu/libpython3.10.so.1.0 /usr/lib/x86_64-linux-gnu/libpython3.8.so.1.0`

**Note: apparently this is not recommended for some reason as libpython3.8 and libpython3.10 are not compatible, but was the only thing I could get to work not being that familiar with ubuntu and compilers** 

I finally thought I could give you a break now Stephan 😂, sorry. I now have a connection timed out when trying to use the GDB port 3333 to connect to the NEORV32:

![image](https://user-images.githubusercontent.com/100368901/219102406-958ef043-47c4-44e7-aca4-3e7d187c2898.png)

What I have done is first opened Vivado and downloaded the bitstream of the on chip debugger test setup to the FPGA. Then used TeraTerm to send the demo blink executable binary and it begins executing that.
Next I open the Windows command prompt and openup OpenOCD. This halts the core and listens on the 3 ports (gdb, tcl and telnet)
Then I open Ubuntu and set path and run the clean_all all command on the demo_blink_led directory. Then open GDB, type in the command in the user guide and it fails to connect.

Any ideas as to what seems to be wrong now that the GDB port can't connect?

**Reply by stnolting** [reply]
> Finally I think it now seems to be working 😅😅😅 :

Great to hear! 🚀

> For anyone facing the same issue, I had to install Python 3.8 on Ubuntu then run the following command:

Thanks for the hint!

> I finally thought I could give you a break now Stephan 😂, sorry. I now have a connection timed out when trying to use the GDB port 3333 to connect to the NEORV32:

Your setup procedure sound fine, so there must be a problem with GDB itself. Can you explain your OS setup? openOCD is running on **native** Windows, but GDB is running on ... a virtual machine under Windows? The WSL? Or a different machine (in the same network)?

Does openOCD issue any notes/warnings during your connection attempt?

**Reply by DS-567** [reply]
> Your setup procedure sound fine, so there must be a problem with GDB itself. Can you explain your OS setup? openOCD is running on native Windows, but GDB is running on ... a virtual machine under Windows? The WSL? Or a different machine (in the same network)?

So yes OpenOCD is used on native Windows command prompt. It seems to be working ok except for the libusb() not found error. 
GDB is running on Ubuntu that I think is running off Windows (not really sure if this is actually what I have but I think it is, I open Ubuntu from the Windows search bar). Both run on the same machine (laptop). 

> Does openOCD issue any notes/warnings during your connection attempt?

This is what happens when I run OpenOCD (see image below). When I try to connect to OpenOCD with GDB, nothing happens on OpenOCD. It remains the same in the picture below.

![image](https://user-images.githubusercontent.com/100368901/219120868-3aea56d9-75ab-477c-ac3c-73b2d4dcb94e.png)

Could it be anything to do with the JTAG reset input of the NEORV32? I have connected it directly to the FTDI board and not pulled it high.

**Reply by stnolting** [reply]
> So yes OpenOCD is used on native Windows command prompt. It seems to be working ok except for the libusb() not found error.

:+1:

> GDB is running on Ubuntu that I think is running off Windows (not really sure if this is actually what I have but I think it is, I open Ubuntu from the Windows search bar). Both run on the same machine (laptop).

This sounds like WSL (Windows Subsystem for Linux). I also using that, but I still use the "old" version (WSL 1 I think). If you are running the new WSL version, the system might have problems to resolve "localhost" (I have seen that before). Can you try `target extended-remote 0.0.0.0:3333` instead of `target extended-remote localhost:3333`?

> Could it be anything to do with the JTAG reset input of the NEORV32? I have connected it directly to the FTDI board and not pulled it high.

I think this is just a problem with GBD. Or more precisely: a problem with the underlying operating system.

**Reply by DS-567** [reply]
> Can you try target extended-remote 0.0.0.0:3333 instead of target extended-remote localhost:3333?

Unfortunately still no connection to NEORV32:

![image](https://user-images.githubusercontent.com/100368901/219123605-e4130757-c654-4006-b639-52256a5ba0e8.png)

**Reply by stnolting** [reply]
Hmm...

Let's check the connection first. Please try to connect to openOCD (from your Ubuntu) using telnet:

```bash
$ telnet localhost 4444
```

or

```bash
$ telnet 0.0.0.0 4444
```

Your Ubuntu terminal should show

```bash
$ telnet localhost 4444
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Open On-Chip Debugger
```

and openOCD should also issue a note:

```bash
Info : accepting 'telnet' connection on tcp/4444
```

**Reply by DS-567** [reply]
> Let's check the connection first. Please try to connect to openOCD (from your Ubuntu) using telnet:

Both of them commands don't seem to work either (I assume you mean to not send these commands through GDB):

![image](https://user-images.githubusercontent.com/100368901/219125866-b887289a-abe0-4ecf-84a3-348d751bdc26.png)

This was done on a separate Ubuntu terminal than the one running the GDB if that matters.

**Reply by stnolting** [reply]
> I assume you mean to not send these commands through GDB

Correct ;)

Ok, so now we know that this is a problem with Windows/WSL2... (at least something! :wink:)

Seems like Windows and WSL do not share "localhost" so you need to explicitly provide your machine's local ip.

You can find the local ip of your machine by running `ipconfig` in a Windows command console. Basically, it should be one of these:

`127.0.0.1` (localhost)
`192.168.0.x`

So please try that ip with telnet.

**Reply by DS-567** [reply]
Just tried `telnet 192.168.x.x 4444` which is my local IP of laptop, from separate Ubuntu terminal from GDB. But after a few minutes of trying 192.168.x.x :

`telnet: Unable to connect to remote host: Connection timed out`

Also tried typical localhost IP:

```
david@DESKTOP-3UBE0P7:/$ telnet 127.0.0.1 4444
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection refused
```

**Reply by stnolting** [reply]
Oh damn! 😅

I am updating my system to WSL2 right now. Let's see if I'll get the same problems...

**Reply by DS-567** [reply]
I should send you some German beer tokens for all this help! 😆😆

This is likely a silly mistake I am making, which if so, I apologise in advance 😄.

**Reply by stnolting** [reply]
It's not your fault, it is a (quite common) problem with WSL2... However, I would take the beer tokens anyway 😅

Now, that I am also on WSL2, I have the same problem... :(

Would be downgrading to WSL1 an option for you?

**Reply by DS-567** [reply]
> Now, that I am also on WSL2, I have the same problem... :(

Ah ok, no problem. So it was an operating system issue then.

> Would be downgrading to WSL1 an option for you?

Yeah if you think that is the easiest (best) solution, I will do that. I only really use Ubuntu at the moment for NEORV32 work.

**Reply by stnolting** [reply]
Ok, great.

So, open a powershell (hit the Windows key and enter Powershell; a blue window should open).
Then enter `wsl -l -v`. This will show the available WSL distros (`-l` = list, `-v` = verbose). Please post the result here (no secrets there :wink:).

This is what I get:
![grafik](https://user-images.githubusercontent.com/22944758/219138372-7af2b30d-e1f1-4583-ab2d-49d0f84af7d1.png)

**Reply by DS-567** [reply]
Here you are: 

![image](https://user-images.githubusercontent.com/100368901/219139340-e138cfa4-a542-4a56-9a8b-df49a09002e6.png)

**Reply by stnolting** [reply]
I think this looks good! Now downgrade to version 1 by running `wsl --set-version Ubuntu 1` in the powershell. This will take some minutes.

**Reply by stnolting** [reply]
And close all Ubuntu consoles before that.

**Reply by DS-567** [reply]
👍

In progress..

**Reply by DS-567** [reply]
Ok, conversion is complete. I have checked and Ubuntu is now version 1.

Shall I try the same process again?

**Reply by stnolting** [reply]
Oh, mine is still in progress 😅
Yes, close the Powershell window, open a new Ubuntu console window and try `telnet localhost 4444` again. Of course, openOCD has to be running in a Windows console.

**Reply by DS-567** [reply]
🥳🥳🥳🥳🥳🥳

![image](https://user-images.githubusercontent.com/100368901/219143090-12726040-051e-48b9-9b09-040874e258a4.png)

**Reply by stnolting** [reply]
Wohoo!!! 🎉

Close that window, open a new Ubuntu window and then run GDB and connect to openOCD: `target extended-remote localhost:3333`

**Reply by DS-567** [reply]
Yeah I got connected perfectly that time 👍. Now I just need to figure out how to use the de-bugger 😂.

Thank you very much Stephan for all your help with this!

**Reply by stnolting** [reply]
You're welcome! I'm really happy that it works now. Finally time for a beer! 😄

You can find a minimal NEORV32+GDB tutorial over here: https://stnolting.github.io/neorv32/ug/#_debugging_using_the_on_chip_debugger

**Reply by DS-567** [reply]
Yeah it works brilliant!

> You're welcome! I'm really happy that it works now. Finally time for a beer! 😄

I think you need more than one beer for putting up with my terminal skills 😂.
